pkg/operator/oauthapiencryptioncontroller/oauthapiencryptioncontroller.go

```diff
@@ -0,0 +1,128 @@
+//         - it will start synchronisation
+//
+// case 2: if the secret exists and it is annotated
+//         - it will simply start synchronisation
+//
+// case 3: no-op: when the secret exits but it doesn't have the annotation - that means it was created by CAO in 4.6 and this is downgrade
+// case 4: no-op: when the secret doesn't exist and encryption is off
+//
+// drawbacks:
+// - it will not recover when the annotation was manually removed by a user,
+//   to recover we would have to put a value in the annotation instead and coordinate OAS-A and CAO but then we would have to remember to add it in CAO (4.6) as well
+func (c *oauthAPIServerController) sync(ctx context.Context, controllerContext factory.SyncContext) error {
+	openshiftAPIServerEncryptionCfg, err := c.secretLister.Get(fmt.Sprintf("%s-%s", encryptionconfig.EncryptionConfSecretName, c.targetNamespace))
+	if apierrors.IsNotFound(err) {
+		return nil // case 4: encryption off
+	}
+	if err != nil {
+		return err
+	}
+
+	oauthAPIServerEncryptionCfg, err := c.secretLister.Get(fmt.Sprintf("%s-%s", encryptionconfig.EncryptionConfSecretName, c.oauthAPIServerTargetNamespace))
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	if apierrors.IsNotFound(err) {
+		// case 1: create with annotation and start synchronization
+		source := openshiftAPIServerEncryptionCfg.DeepCopy()
+		source.Name = fmt.Sprintf("%s-%s", encryptionconfig.EncryptionConfSecretName, c.oauthAPIServerTargetNamespace)
+		source.ResourceVersion = ""
+		source.OwnerReferences = []metav1.OwnerReference{}
+		if source.Annotations == nil {
+			source.Annotations = map[string]string{}
+		}
+		source.Annotations[encryptionConfigManagedBy] = encryptionConfigManagedByValue
+		if _, _, err = resourceapply.ApplySecret(c.secretClient, controllerContext.Recorder(), source); err != nil {
+			return err
+		}
+		return c.startSecretSynchronization(controllerContext.Recorder())
+	}
+	if _, exist := oauthAPIServerEncryptionCfg.Annotations[encryptionConfigManagedBy]; exist {
+		// case 2: exists and it is annotated
+		return c.startSecretSynchronization(controllerContext.Recorder())
+	}
+
+	// case 3: no-op the secret is managed by CAO
+	return nil
+}
+
+func (c *oauthAPIServerController) startSecretSynchronization(recorder events.Recorder) error {
+	if c.synchronizationEnabled {
+		return nil
+	}
+
+	if err := c.resourceSyncer.SyncSecret(
+		resourcesynccontroller.ResourceLocation{Namespace: operatorclient.GlobalMachineSpecifiedConfigNamespace, Name: fmt.Sprintf("%s-%s", encryptionconfig.EncryptionConfSecretName, c.oauthAPIServerTargetNamespace)},
+		resourcesynccontroller.ResourceLocation{Namespace: operatorclient.GlobalMachineSpecifiedConfigNamespace, Name: fmt.Sprintf("%s-%s", encryptionconfig.EncryptionConfSecretName, c.targetNamespace)},
+	); err != nil {
+		return err
+	}
+	c.synchronizationEnabled = true
```

Comment:
why do we need this state? It's surprising.

Reply:
I don't want to call `c.resourceSyncer.SyncSecret` multiple times - only on startup or when encryption has been turned on.

Reply:
I thought we have to call it continuously

Reply:
`c.resourceSyncer` is a controller.

Reply:
I don't like state in controllers which is changed during the controller runtime. It's a bad pattern leading to bugs IMO.